# Workflow 1, Module 3, Question 2

## What proteins produce agent [x]?

Let's run Q1 and use the endogenous output of that.

### Expand service

In [1]:
robokop_server = 'robokop.renci.org'

In [2]:
import requests
import pandas as pd

In [3]:
def expand(type1,identifier,type2,rebuild=None,csv=None,predicate=None):
    url=f'http://{robokop_server}:80/api/simple/expand/{type1}/{identifier}/{type2}'
    params = {'rebuild': rebuild, 
              'csv'    : csv,
              'predicate': predicate} 
    params = { k:v for k,v in params.items() if v is not None }
    response = requests.get(url,params=params)
    print( f'Return Status: {response.status_code}' )
    if response.status_code == 200:
        return response.json()
    return []

In [4]:
def parse_answer(returnanswer):
    nodes = [answer['nodes'][1] for answer in returnanswer['answers']]
    edges = [answer['edges'][0] for answer in returnanswer['answers']]
    answers = [ {"result_id": node["id"], 
                 "result_name": node["name"] if 'name' in node else node['id'], 
                 "type": edge["type"],
                 "source": edge['edge_source']}
              for node,edge in zip(nodes,edges)]
    return pd.DataFrame(answers)

### Question 1

In [5]:
input_identifier='MONDO:0004979'
type1 = 'disease'
type2 = 'chemical_substance'

In [6]:
result = expand(type1,input_identifier,type2)
result_frame = parse_answer(result)
endogenous = result_frame[ result_frame['type'] == 'related_to' ]
endogenous

Return Status: 200


result_id                                result_name  \
23      CHEBI:68447                      1,3-dimethyluric acid   
38      CHEBI:28946                                Theobromine   
49      CHEBI:25858                               Paraxanthine   
50      CHEBI:68449                      1,7-Dimethyluric acid   
51     CHEBI:691622                   1,3,7-trimethyluric acid   
96      CHEBI:27732                                   Caffeine   
118     CHEBI:62205                           3-methylxanthine   
238     CHEBI:68444                           1-methylxanthine   
239     CHEBI:27595                                  11b-PGF2a   
243     CHEBI:62207                       3-methyl-7H-xanthine   
244     CHEBI:28209  5-hydroxy-6,8,11,14-eicosatetraenoic acid   
490  PUBCHEM:119226                           3-Chlorotyrosine   
509  PUBCHEM:273429                          Lactosylceramides   

                         source        type  
23   hmdb.metabolite_to_disease  related_to  
38   hmdb.metabolite_to_disease  related_to  
49   hmdb.metabolite_to_disease  related_to  
50   hmdb.metabolite_to_disease  related_to  
51   hmdb.metabolite_to_disease  related_to  
96   hmdb.metabolite_to_disease  related_to  
118  hmdb.metabolite_to_disease  related_to  
238  hmdb.metabolite_to_disease  related_to  
239  hmdb.metabolite_to_disease  related_to  
243  hmdb.metabolite_to_disease  related_to  
244  hmdb.metabolite_to_disease  related_to  
490  hmdb.disease_to_metabolite  related_to  
509  hmdb.disease_to_metabolite  related_to

### Question 2

In [7]:
identifiers = list(endogenous['result_id'])

In [8]:
genes = {}
for ident in identifiers:
    print(ident)
    genes[ident] = expand('chemical_substance',ident,'gene',rebuild=True)

CHEBI:68447
Return Status: 200
CHEBI:28946
Return Status: 200
CHEBI:25858
Return Status: 200
CHEBI:68449
Return Status: 200
CHEBI:691622
Return Status: 200
CHEBI:27732
Return Status: 200
CHEBI:62205
Return Status: 200
CHEBI:68444
Return Status: 200
CHEBI:27595
Return Status: 200
CHEBI:62207
Return Status: 200
CHEBI:28209
Return Status: 200
PUBCHEM:119226
Return Status: 200
PUBCHEM:273429
Return Status: 200


In [9]:
results = []
for chemical in genes:
    if 'answers' in genes[chemical]:
        frame = parse_answer(genes[chemical])
        frame['Chemical'] = chemical
        results.append(frame)
all_results = pd.concat(results)
all_results

result_id     result_name                     source  \
0         HGNC:2596          CYP1A2           ctd.drug_to_gene   
1         HGNC:2631          CYP2E1           ctd.drug_to_gene   
2         HGNC:8781           PDE4B  hmdb.metabolite_to_enzyme   
3        HGNC:12805             XDH  hmdb.metabolite_to_enzyme   
4          HGNC:262          ADORA1  hmdb.metabolite_to_enzyme   
5          HGNC:263         ADORA2A  hmdb.metabolite_to_enzyme   
6    EC:1.14.13.178  EC:1.14.13.178   kegg.chemical_get_enzyme   
7      EC:2.1.1.159    EC:2.1.1.159   kegg.chemical_get_enzyme   
8      EC:1.13.12.-    EC:1.13.12.-   kegg.chemical_get_enzyme   
9       EC:1.17.3.2     EC:1.17.3.2   kegg.chemical_get_enzyme   
10   EC:1.14.13.179  EC:1.14.13.179   kegg.chemical_get_enzyme   
11     EC:2.1.1.160    EC:2.1.1.160   kegg.chemical_get_enzyme   
0         HGNC:2596          CYP1A2   kegg.chemical_get_enzyme   
1         HGNC:2610          CYP2A6  hmdb.enzyme_to_metabolite   
2         HGNC:7646            NAT2   kegg.chemical_get_enzyme   
3        HGNC:12805             XDH  hmdb.enzyme_to_metabolite   
4         HGNC:2622          CYP2C8  hmdb.enzyme_to_metabolite   
5         HGNC:5209         HSD11B2           ctd.drug_to_gene   
6         HGNC:2638          CYP3A5  hmdb.enzyme_to_metabolite   
7         HGNC:2615          CYP2B6  hmdb.enzyme_to_metabolite   
8         HGNC:2621         CYP2C19  hmdb.enzyme_to_metabolite   
9         HGNC:2623          CYP2C9  hmdb.enzyme_to_metabolite   
10        HGNC:2631          CYP2E1  hmdb.enzyme_to_metabolite   
11        HGNC:2625          CYP2D6  hmdb.enzyme_to_metabolite   
12        HGNC:2637          CYP3A4  hmdb.enzyme_to_metabolite   
13        HGNC:2595          CYP1A1  hmdb.enzyme_to_metabolite   
14        HGNC:6018             IL6  ctd.drug_to_gene_extended   
15       HGNC:11892             TNF           ctd.gene_to_drug   
16       HGNC:20244          CYP4X1  hmdb.enzyme_to_metabolite   
17        HGNC:2632          CYP2F1  hmdb.enzyme_to_metabolite   
..              ...             ...                        ...   
132     EC:1.17.5.2     EC:1.17.5.2   kegg.chemical_get_enzyme   
133    EC:2.1.1.160    EC:2.1.1.160   kegg.chemical_get_enzyme   
134    EC:1.13.12.-    EC:1.13.12.-   kegg.chemical_get_enzyme   
0         HGNC:2596          CYP1A2           ctd.drug_to_gene   
1        HGNC:12805             XDH           ctd.drug_to_gene   
2    EC:1.14.13.179  EC:1.14.13.179   kegg.chemical_get_enzyme   
3       EC:1.17.3.2     EC:1.17.3.2   kegg.chemical_get_enzyme   
4    EC:1.14.13.178  EC:1.14.13.178   kegg.chemical_get_enzyme   
5      EC:1.13.12.-    EC:1.13.12.-   kegg.chemical_get_enzyme   
0          HGNC:386          AKR1C3  hmdb.metabolite_to_enzyme   
0      EC:1.13.12.-    EC:1.13.12.-   kegg.chemical_get_enzyme   
1    EC:1.14.13.179  EC:1.14.13.179   kegg.chemical_get_enzyme   
2    EC:1.14.13.178  EC:1.14.13.178   kegg.chemical_get_enzyme   
0         HGNC:6877           MAPK3           ctd.drug_to_gene   
1         HGNC:7794           NFKB1  ctd.drug_to_gene_extended   
2         HGNC:7940            NPPB           ctd.drug_to_gene   
3         HGNC:3402           EPHX2  ctd.drug_to_gene_extended   
4         HGNC:6871           MAPK1           ctd.drug_to_gene   
5         HGNC:9955            RELA  ctd.drug_to_gene_extended   
6         HGNC:4556            GPX4  hmdb.metabolite_to_enzyme   
7         HGNC:4558            GPX6  hmdb.metabolite_to_enzyme   
8         HGNC:4559            GPX7  hmdb.metabolite_to_enzyme   
9         HGNC:4555            GPX3  hmdb.metabolite_to_enzyme   
10        HGNC:4557            GPX5  hmdb.metabolite_to_enzyme   
11        HGNC:4554            GPX2  hmdb.metabolite_to_enzyme   
12        HGNC:4553            GPX1  hmdb.metabolite_to_enzyme   
13        HGNC:6893            MAPT       pharos.drug_get_gene   
14         HGNC:435           ALOX5           ctd.drug_to_gene   
15       HGNC:33100            GPX8   kegg.chemica

Are there any predicates indicating a production?

In [10]:
all_results['type'].value_counts()

interacts_with                            64
decreases_expression_of                   48
increases_expression_of                   27
negatively_regulates__entity_to_entity    20
positively_regulates__entity_to_entity    12
directly_interacts_with                   11
affects_expression_of                      9
increases_activity_of                      6
increases_molecular_modification_of        5
decreases_activity_of                      4
increases_metabolic_processing_of          3
increases_synthesis_of                     3
affects_response_to                        2
decreases_molecular_modification_of        1
affects_metabolic_processing_of            1
increases_degradation_of                   1
affects_activity_of                        1
Name: type, dtype: int64

In [17]:
all_results[all_results['type']=='positively_regulates__entity_to_entity']

result_id     result_name                    source  \
30   EC:1.14.13.179  EC:1.14.13.179  kegg.chemical_get_enzyme   
115    EC:2.1.1.160    EC:2.1.1.160  kegg.chemical_get_enzyme   

                                       type     Chemical  
30   positively_regulates__entity_to_entity  CHEBI:25858  
115  positively_regulates__entity_to_entity  CHEBI:27732

We're finding enzymes involved in creating the endogenous chemicals. But how come our genes are in EC format, rather than being translated to our normal HGNC?   It's because these are non-human enzymes and at the moment we're not synonymizing non-human genes. In fact, for the most part, we don't accept non-human genes, but we're not checking for that in the KEGG client.

In [11]:
all_results[all_results['type']=='increases_synthesis_of']

result_id result_name            source                    type  \
1   HGNC:2610      CYP2A6  ctd.drug_to_gene  increases_synthesis_of   
14   HGNC:435       ALOX5  ctd.drug_to_gene  increases_synthesis_of   
0     HGNC:40       ABCB1  ctd.drug_to_gene  increases_synthesis_of   

          Chemical  
1      CHEBI:68449  
14     CHEBI:28209  
0   PUBCHEM:273429

So here's an example of a metabolite (`PUBCHEM:273429`) that is increased by the action of a gene `ABC1`

### Cystic Fibrosis

In [12]:
cf = 'MONDO:0009061'
result = expand(type1,cf,type2)
result_frame = parse_answer(result)
endogenous = result_frame[ result_frame['type'] == 'related_to' ]
endogenous

Return Status: 200


result_id                   result_name  \
4            CHEBI:31835                 methylparaben   
5            CHEBI:16325              Lithocholic Acid   
6            CHEBI:30805               Dodecanoic acid   
7            CHEBI:30813                   Capric acid   
8            CHEBI:17201                Glycyl-glycine   
10           CHEBI:28621               Triethanolamine   
12           CHEBI:16755         Chenodeoxycholic Acid   
13           CHEBI:28834              Deoxycholic Acid   
14            CHEBI:6650                    Malic acid   
15           CHEBI:16359                   Cholic Acid   
27           CHEBI:16919                      Creatine   
32           CHEBI:16737                    Creatinine   
36           CHEBI:17196                  L-Asparagine   
37           CHEBI:17712                      Xanthine   
38            CHEBI:8207                   Pilocarpine   
39           CHEBI:28911                     Cobalamin   
40           CHEBI:32063                 propylparaben   
41           CHEBI:17243  Monoethylhexyl phthalic acid   
42           CHEBI:49059                 Threonic acid   
65       PUBCHEM:3952518         Prolyl-Hydroxyproline   
67        PUBCHEM:417360              Glycyl-Histidine   
69      PUBCHEM:23617285              Glycocholic acid   
72  CHEMBL:CHEMBL2004044                 Glycyl-Serine   

                        source        type  
4   hmdb.metabolite_to_disease  related_to  
5   hmdb.metabolite_to_disease  related_to  
6   hmdb.metabolite_to_disease  related_to  
7   hmdb.metabolite_to_disease  related_to  
8   hmdb.metabolite_to_disease  related_to  
10  hmdb.metabolite_to_disease  related_to  
12  hmdb.metabolite_to_disease  related_to  
13  hmdb.metabolite_to_disease  related_to  
14  hmdb.metabolite_to_disease  related_to  
15  hmdb.metabolite_to_disease  related_to  
27  hmdb.metabolite_to_disease  related_to  
32  hmdb.metabolite_to_disease  related_to  
36  hmdb.metabolite_to_disease  related_to  
37  hmdb.metabolite_to_disease  related_to  
38  hmdb.metabolite_to_disease  related_to  
39  hmdb.metabolite_to_disease  related_to  
40  hmdb.metabolite_to_disease  related_to  
41  hmdb.metabolite_to_disease  related_to  
42  hmdb.metabolite_to_disease  related_to  
65  hmdb.disease_to_metabolite  related_to  
67  hmdb.disease_to_metabolite  related_to  
69  hmdb.disease_to_metabolite  related_to  
72  hmdb.disease_to_metabolite  related_to

In [13]:
identifiers = list(endogenous['result_id'])
genes = {}
for ident in identifiers:
    print(ident)
    genes[ident] = expand('chemical_substance',ident,'gene',rebuild=True)

CHEBI:31835
Return Status: 200
CHEBI:16325
Return Status: 200
CHEBI:30805
Return Status: 200
CHEBI:30813
Return Status: 200
CHEBI:17201
Return Status: 200
CHEBI:28621
Return Status: 200
CHEBI:16755
Return Status: 200
CHEBI:28834
Return Status: 200
CHEBI:6650
Return Status: 200
CHEBI:16359
Return Status: 200
CHEBI:16919
Return Status: 200
CHEBI:16737
Return Status: 200
CHEBI:17196
Return Status: 200
CHEBI:17712
Return Status: 200
CHEBI:8207
Return Status: 200
CHEBI:28911
Return Status: 200
CHEBI:32063
Return Status: 200
CHEBI:17243
Return Status: 200
CHEBI:49059
Return Status: 200
PUBCHEM:3952518
Return Status: 200
PUBCHEM:417360
Return Status: 200
PUBCHEM:23617285
Return Status: 200
CHEMBL:CHEMBL2004044
Return Status: 200


In [14]:
results = []
for chemical in genes:
    if 'answers' in genes[chemical]:
        frame = parse_answer(genes[chemical])
        frame['Chemical'] = chemical
        results.append(frame)
all_results = pd.concat(results)
all_results['type'].value_counts()

increases_expression_of                   203
interacts_with                            119
decreases_expression_of                    87
directly_interacts_with                    43
negatively_regulates__entity_to_entity     24
positively_regulates__entity_to_entity     22
affects_response_to                        20
increases_activity_of                      18
increases_molecular_modification_of        12
affects_expression_of                       9
decreases_activity_of                       8
increases_degradation_of                    5
increases_metabolic_processing_of           2
affects_molecular_modification_of           2
decreases_abundance_of                      2
increases_secretion_of                      2
affects_localization_of                     2
increases_localization_of                   1
decreases_molecular_modification_of         1
increases_synthesis_of                      1
increases_uptake_of                         1
decreases_metabolic_processing_of 

In [15]:
all_results[all_results['type']=='positively_regulates__entity_to_entity']

result_id     result_name                    source  \
78      EC:2.8.3.25     EC:2.8.3.25  kegg.chemical_get_enzyme   
9       EC:3.1.2.21     EC:3.1.2.21  kegg.chemical_get_enzyme   
0       EC:3.5.2.13     EC:3.5.2.13  kegg.chemical_get_enzyme   
118     EC:3.5.1.24     EC:3.5.1.24  kegg.chemical_get_enzyme   
119      EC:6.2.1.7      EC:6.2.1.7  kegg.chemical_get_enzyme   
61      EC:3.5.1.24     EC:3.5.1.24  kegg.chemical_get_enzyme   
62      EC:2.8.3.25     EC:2.8.3.25  kegg.chemical_get_enzyme   
16       HGNC:15919           ACOT8  kegg.chemical_get_enzyme   
48      EC:3.5.1.24     EC:3.5.1.24  kegg.chemical_get_enzyme   
14       EC:3.9.1.1      EC:3.9.1.1  kegg.chemical_get_enzyme   
15      EC:3.5.2.10     EC:3.5.2.10  kegg.chemical_get_enzyme   
16       EC:2.1.1.2      EC:2.1.1.2  kegg.chemical_get_enzyme   
1          HGNC:753            ASNS  kegg.chemical_get_enzyme   
14       EC:6.3.1.4      EC:6.3.1.4  kegg.chemical_get_enzyme   
15       EC:6.3.1.1      EC:6.3.1.1  kegg.chemical_get_enzyme   
18      EC:3.2.2.11     EC:3.2.2.11  kegg.chemical_get_enzyme   
7        EC:3.2.2.1      EC:3.2.2.1  kegg.chemical_get_enzyme   
8    EC:1.14.13.178  EC:1.14.13.178  kegg.chemical_get_enzyme   
9    EC:1.14.13.128  EC:1.14.13.128  kegg.chemical_get_enzyme   
11   EC:1.14.13.179  EC:1.14.13.179  kegg.chemical_get_enzyme   
12       EC:3.5.4.3      EC:3.5.4.3  kegg.chemical_get_enzyme   
0       EC:3.1.1.60     EC:3.1.1.60  kegg.chemical_get_enzyme   

                                       type     Chemical  
78   positively_regulates__entity_to_entity  CHEBI:16325  
9    positively_regulates__entity_to_entity  CHEBI:30813  
0    positively_regulates__entity_to_entity  CHEBI:17201  
118  positively_regulates__entity_to_entity  CHEBI:16755  
119  positively_regulates__entity_to_entity  CHEBI:16755  
61   positively_regulates__entity_to_entity  CHEBI:28834  
62   positively_regulates__entity_to_entity  CHEBI:28834  
16   positively_regulates__entity_to_entity  CHEBI:16359  
48   positively_regulates__entity_to_entity  CHEBI:16359  
14   positively_regulates__entity_to_entity  CHEBI:16919  
15   positively_regulates__entity_to_entity  CHEBI:16919  
16   positively_regulates__entity_to_entity  CHEBI:16919  
1    positively_regulates__entity_to_entity  CHEBI:17196  
14   positively_regulates__entity_to_entity  CHEBI:17196  
15   positively_regulates__entity_to_entity  CHEBI:17196  
18   positively_regulates__entity_to_entity  CHEBI:17196  
7    positively_regulates__entity_to_entity  CHEBI:17712  
8    positively_regulates__entity_to_entity  CHEBI:17712  
9    positively_regulates__entity_to_entity  CHEBI:17712  
11   positively_regulates__entity_to_entity  CHEBI:17712  
12   positively_regulates__entity_to_entity  CHEBI:17712  
0    positively_regulates__entity_to_entity  CHEBI:17243

In [16]:
all_results[all_results['type']=='increases_synthesis_of']

result_id result_name            source                    type  \
13  HGNC:2768         DEK  ctd.drug_to_gene  increases_synthesis_of   

       Chemical  
13  CHEBI:16919